File: AquariumApi.Models/Device/DeviceSchedule.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Threading.Tasks;

namespace AquariumApi.Models
{
    [Table("tblDeviceSchedule")]
    public class DeviceSchedule
    {
        public int? Id { get; set; }
        public string Name { get; set; }
        public int DeviceId { get; set; }
        public DateTime StartTime { get; set; }
        public List<DeviceScheduleTask> Tasks { get; set; }
        public bool Repeat { get; set; }
        public int RepeatInterval { get; set; }
        public DateTime RepeatEndTime { get; set; }
        public string DateConditions { get; set; }
    }
}

File: AquariumApi.DeviceApi/ClientApp/src/app/modules/SharedDeviceModule/models/DeviceSchedule.ts
import { DeviceConfiguration } from './DeviceConfiguration'
import { DeviceScheduleTask } from './DeviceScheduleTask'

export class DeviceSchedule
{
  id: number //= Math.floor(Math.random() * 100);
  name: string
  deviceId: number
  startTime: string

  tasks: DeviceScheduleTask[] = []

  repeat: boolean
  repeatInterval: number
  repeatEndTime: string
  dateConditions: string

  static expandSchedule(schedule: DeviceSchedule, device: DeviceConfiguration):DeviceSchedule {
    var schedule = {...schedule}
    if(!schedule.tasks)
      schedule.tasks = [];
    schedule.tasks = schedule.tasks.map(t => {
      return DeviceScheduleTask.expandTask(t,device);
    })
    return schedule;
  }
}
